lib/runCall.js

```diff
@@ -83,26 +83,29 @@ module.exports = function (opts, cb) {
     }
   }
 
-  function subTxValue () {
+  function subTxValue (cb) {
     if (delegatecall) {
+      cb()
       return
     }
-    account.balance = new BN(account.balance).sub(txValue)
-    stateManager.cache.put(caller, account)
+    var newBalance = new BN(account.balance).sub(txValue)
+    account.balance = newBalance
+    stateManager.putAccountBalance(ethUtil.toBuffer(caller), newBalance, cb)
   }
 
-  function addTxValue () {
+  function addTxValue (cb) {
     if (delegatecall) {
+      cb()
       return
     }
     // add the amount sent to the `to` account
-    toAccount.balance = new BN(toAccount.balance).add(txValue)
-    stateManager.cache.put(toAddress, toAccount)
-    stateManager.touched.push(toAddress)
+    var newBalance = new BN(toAccount.balance).add(txValue)
+    toAccount.balance = newBalance
+    // _putAccount as the nonce may have changed for contract creation
+    stateManager._putAccount(ethUtil.toBuffer(toAddress), toAccount, cb)
```

Comment:
The functions with an underscore would be internal functions of the `StateManager` and should not be called from outside. (It may be the case at the moment in some places but that should be fixed.)

Reply:
Agreed. The only way of achieving this write in the current codebase is to either manipulate the cache directly (which I feel should also be a no-no) or to call `_putAccount`. My preference would be to make `_putAccount` public (rename to `putAccount`) but I didn't want to change the interface in this way without checking first. Would you be ok with this?